sdk/python/src/dagger/engine/bin.py

```diff
@@ -113,3 +117,141 @@ async def start(self) -> None:
+        cache_dir = (
+            Path(os.environ.get("XDG_CACHE_HOME", "~/.cache")).expanduser() / "dagger"
+        )
+        cache_dir.mkdir(mode=0o700, parents=True, exist_ok=True)
+
+        os_, arch = get_platform()
+
+        cli_version = self.cfg.host.netloc + self.cfg.host.path
+
+        dagger_cli_bin_path = cache_dir / f"{DAGGER_CLI_BIN_PREFIX}{cli_version}"
+        if os_ == "windows":
+            dagger_cli_bin_path = dagger_cli_bin_path.with_suffix(".exe")
+
+        if not dagger_cli_bin_path.exists():
+            expected_hash = expected_checksum(cli_version)
+            tempfile_args = {
+                "prefix": f"temp-{DAGGER_CLI_BIN_PREFIX}",
+                "dir": cache_dir,
+                "delete": False,
+            }
+            with tempfile.NamedTemporaryFile(**tempfile_args) as tmp_bin:
+
+                def cleanup():
+                    """Remove the tmp_bin on error."""
+                    tmp_bin.close()
+                    os.unlink(tmp_bin.name)
+
+                try:
+                    actual_hash = extract_cli_archive(
+                        cli_version,
+                        tmp_bin,
+                    )
+                    if actual_hash != expected_hash:
+                        raise ProvisionError(
+                            f"Downloaded CLI binary checksum ({actual_hash})"
+                            f"does not match expected checksum ({expected_hash})"
+                        )
+                    # Flake8 Ignores
+                    # F811 -- redefinition of (yet) unused function
+                    # E731 -- assigning a lambda.
+                    cleanup = lambda: None  # noqa: F811,E731
+                finally:
+                    cleanup()
+
+                tmp_bin.close()
+                tmp_bin_path = Path(tmp_bin.name)
+                tmp_bin_path.chmod(0o700)
+
+                dagger_cli_bin_path = tmp_bin_path.rename(dagger_cli_bin_path)
+
+        # garbage collection of old engine_session binaries
+        for bin in cache_dir.glob(f"{DAGGER_CLI_BIN_PREFIX}*"):
+            if bin != dagger_cli_bin_path:
+                bin.unlink()
+
+        self._start(
+            [dagger_cli_bin_path, "session"],
+        )
+
+
+def get_platform() -> tuple[str, str]:
+    normalized_arch = {
+        "x86_64": "amd64",
+        "aarch64": "arm64",
+    }
+    uname = platform.uname()
+    os_name = uname.system.lower()
+    arch = uname.machine.lower()
+    arch = normalized_arch.get(arch, arch)
+    return os_name, arch
+
+
+def cli_archive_name(cli_version: str):
+    os_, arch = get_platform()
+    return f"dagger_v{cli_version}_{os_}_{arch}.tar.gz"
+
+
+def cli_archive_url(cli_version: str):
+    return (
+        f"{CLI_SCHEME}://{CLI_HOST}/dagger/releases/{cli_version}"
+        f"/{cli_archive_name(cli_version)}"
+    )
+
+
+def cli_checksum_url(cli_version: str):
+    return f"{CLI_SCHEME}https://{CLI_HOST}/dagger/releases/{cli_version}/checksums.txt"
```

Comment:
Typo: 

```suggestion
    return f"{CLI_SCHEME}://{CLI_HOST}/dagger/releases/{cli_version}/checksums.txt"
```

Reply:
Oops good catch, thanks, I missed this because I didn't realize we were skipping this test in mage (`-m not provision`), got rid of that now so the test runs and passes